contracts/Marketplace.sol

```diff
@@ -1,74 +1,209 @@
+
+contract Marketplace is Ownable, Pausable {
+  // ------------------------------------------------------
+  // Structures
+  // ------------------------------------------------------
+
+  struct Service {
+    address payable owner;
+    bytes sid;
+    uint price;
+
+    Version[] versions;
+    Payment[] payments;
+  }
+
+  struct Version {
+    bytes20 hash;
+    bytes url;
+  }
+
+  struct Payment {
+    address purchaser;
+  }
+
+  // ------------------------------------------------------
+  // Events
+  // ------------------------------------------------------
+
+  // TODO: indexed on bytes20 transform it as bytes32
+  // TODO: indexed on bytes transform it to something else
+  // TODO: apply same logic for event data and index
+  event ServiceCreated(uint serviceIndex, bytes sid, address indexed owner, uint price);
+
+  event ServiceOwnershipTransferred(uint serviceIndex, bytes sid, address indexed previousOwner, address indexed newOwner);
+
+  event ServicePriceChanged(uint serviceIndex, bytes sid, uint previousPrice, uint newPrice);
+
+  event ServiceVersionCreated(uint serviceIndex, bytes20 hash, bytes url);
+
+  event ServicePaid(uint serviceIndex, bytes sid, address indexed purchaser, address indexed seller, uint price);
+
+  // ------------------------------------------------------
+  // State variables
+  // ------------------------------------------------------
+
+  Service[] public services;
+
+  // ------------------------------------------------------
+  // Constructor
+  // ------------------------------------------------------
+
+  constructor() public {}
+
+  // ------------------------------------------------------
+  // Check functions
+  // ------------------------------------------------------
+
+  function checkServiceOwner(Service memory service) private view {
+    require(service.owner == msg.sender, "Service owner is not the same as the sender");
+  }
+
+  // ------------------------------------------------------
+  // Utils functions
+  // ------------------------------------------------------
 
-    event VersionCreated(bytes32 sid, bytes32 id, bytes32 location, uint256 price);
+  function compare(bytes memory a, bytes memory b) internal pure returns (bool) {
+    return keccak256(a) == keccak256(b);
+  }
 
-    constructor(address creator, bytes32 _sid) Ownable() Pausable() public {
-        sid = _sid;
-        transferOwnership(creator);
-        addPauser(creator);
-        renouncePauser();
+  // ------------------------------------------------------
+  // View functions
+  // ------------------------------------------------------
+
+  // TODO: have to create getter for version and payment because services auto-generated getter doesn't return array of struct
+  function getVersion(uint serviceIndex, uint versionIndex) public view returns (bytes20 hash, bytes memory url) {
+    Version memory version = services[serviceIndex].versions[versionIndex];
+    return (version.hash, version.url);
```

Comment:
- services[serviceIndex].versions[versionIndex];
- why return memory and not from storage;
- s/getVersion/getServiceVersion/

?

Reply:
From this article https://medium.com/coinmonks/ethereum-solidity-memory-vs-storage-which-to-use-in-local-functions-72b593c3703a
I understand that created a memory variable that ref a storage copy the data and thus is expensive.
So in this function, i should replace memory by storage because the data already exist in storage, so it will just create a reference.

Reply:
I'm not sure about if it's better to use memory or storage in function parameters.